java/ffi/src/main/java/sleeper/foreign/FFIBridge.java

```diff
@@ -0,0 +1,118 @@
+     * are encouraged to cache results of successful calls to this function.
+     *
+     * @param  clazz       the interface describing the foreign function calls
+     * @param  <T>         interface type containing Java functions stubs
+     * @return             the native call interface
+     * @throws IOException if an error occurs during loading or linking the native library
+     */
+    public static synchronized <T extends ForeignFunctions> T createForeignInterface(Class<T> clazz) throws IOException {
+        try {
+            return extractAndLink(clazz, "compaction");
+        } catch (UnsatisfiedLinkError err) {
+            throw (IOException) new IOException("Could not initialise foreign library bridge", err);
+        }
+    }
+
+    /**
+     * Loads the named library after extracting it from the classpath.
+     *
+     * This function extracts the named library from a JAR on the classpath and attempts to load it
+     * and bind it to the given interface class. The paths in the array {@link LIB_PATHS} are tried
+     * in order. If a library is found at a path, this method will attempt to load it. If no library
+     * is found on the classpath or it can't be loaded (e.g. wrong binary format), the next path
+     * will be tried.
+     *
+     * The library named should be given without platform prefixes, e.g. "foo" will be expanded into
+     * "libfoo.so" or "foo.dll" as appropriate for this platform.
+     *
+     * @param  <T>                  the type of the interface to the native code
+     * @param  clazz                the class of the interface to the native code
+     * @param  libName              the library name to extract without platform prefixes.
+     * @return                      the absolute extracted path, or null if the library couldn't be found
+     * @throws IOException          if an error occured during file extraction
+     * @throws UnsatisfiedLinkError if the library could not be found or loaded
+     */
+    public static <T> T extractAndLink(Class<T> clazz, String libName) throws IOException {
+        // Work through each potential path to see if we can load the library
+        // successfully
+        for (String path : LIB_PATHS) {
+            LOGGER.debug("Attempting to load native library from JAR path {}", path);
+            // Attempt extraction
+            File extractedLib = EXTRACTOR.extractJni(path, libName);
+
+            // If file located, attempt to load
+            if (extractedLib != null) {
+                LOGGER.debug("Extracted file is at {}", extractedLib);
+                try {
+                    return LibraryLoader.create(clazz).failImmediately()
+                            .load(extractedLib.getAbsolutePath());
+                } catch (UnsatisfiedLinkError e) {
+                    // wrong library, try the next path
+                    LOGGER.error("Unable to load native library from " + path, e);
```

Comment:
This looks like it should be warn level rather than error, since it's normal for some libraries to fail?

The path can be parameterised with a placeholder rather than concatenated.